app/models/category.rb

```diff
@@ -153,23 +154,40 @@ def reset_topic_ids_cache
     Category.reset_topic_ids_cache
   end
 
+  @@subcategory_ids = DistributedCache.new('subcategory_ids')
+
   def self.subcategory_ids(category_id)
-    sql = <<~SQL
-        WITH RECURSIVE subcategories AS (
-            SELECT :category_id id, 1 depth
-            UNION
-            SELECT categories.id, (subcategories.depth + 1) depth
-            FROM categories
-            JOIN subcategories ON subcategories.id = categories.parent_category_id
-            WHERE subcategories.depth < :max_category_nesting
+    @@subcategory_ids[category_id] ||=
+      begin
+        sql = <<~SQL
+            WITH RECURSIVE subcategories AS (
+                SELECT :category_id id, 1 depth
+                UNION
+                SELECT categories.id, (subcategories.depth + 1) depth
+                FROM categories
+                JOIN subcategories ON subcategories.id = categories.parent_category_id
+                WHERE subcategories.depth < :max_category_nesting
+            )
+            SELECT id FROM subcategories
+          SQL
+        DB.query_single(
+          sql,
+          category_id: category_id,
+          max_category_nesting: SiteSetting.max_category_nesting
         )
-        SELECT id FROM subcategories
-      SQL
-    DB.query_single(
-      sql,
-      category_id: category_id.to_i,
-      max_category_nesting: SiteSetting.max_category_nesting
-    )
+      end
+  end
+
+  def self.clear_subcategory_ids
+    @@subcategory_ids.clear
+  end
+
+  DiscourseEvent.on(:site_setting_changed) do |name, _, _|
```

Comment:
I think we try to keep these types of things in `config/initializers/014-track-setting-changes.rb`, so that it's easier to reason with.

Also, do we need to clear the cache in any other situations? E.g. restoring a backup?